File: ch5/UVa10229.cpp
// Modular Fibonacci

#include <bits/stdc++.h>
using namespace std;

typedef long long ll;

ll MOD;

const int MAX_N = 2;                             // 2x2 for Fib matrix

struct Matrix { ll mat[MAX_N][MAX_N]; };         // we return a 2D array

ll mod(ll a, ll m) { return ((a%m)+m) % m; }     // ensure positive answer

Matrix matMul(Matrix a, Matrix b) {              // normally O(n^3)
  Matrix ans;                                    // but O(1) as n = 2
  for (int i = 0; i < MAX_N; ++i)
    for (int j = 0; j < MAX_N; ++j)
      ans.mat[i][j] = 0;
  for (int i = 0; i < MAX_N; ++i)
    for (int k = 0; k < MAX_N; ++k) {
      if (a.mat[i][k] == 0) continue;            // optimization
      for (int j = 0; j < MAX_N; ++j) {
        ans.mat[i][j] += mod(a.mat[i][k], MOD) * mod(b.mat[k][j], MOD);
        ans.mat[i][j] = mod(ans.mat[i][j], MOD); // modular arithmetic
      }
    }
  return ans;
}

Matrix matPow(Matrix base, int p) {              // normally O(n^3 log p)
  Matrix ans;                                    // but O(log p) as n = 2
  for (int i = 0; i < MAX_N; ++i)
    for (int j = 0; j < MAX_N; ++j)
      ans.mat[i][j] = (i == j);                  // prepare identity matrix
  while (p) {                                    // iterative D&C version
    if (p&1)                                     // check if p is odd
      ans = matMul(ans, base);                   // update ans
    base = matMul(base, base);                   // square the base
    p >>= 1;                                     // divide p by 2
  }
  return ans;
}

int main() {
  int n, m;
  while (scanf("%d %d", &n, &m) == 2) {
    Matrix ans;                                  // Fibonaccci matrix
    ans.mat[0][0] = 1;  ans.mat[0][1] = 1;
    ans.mat[1][0] = 1;  ans.mat[1][1] = 0;
    MOD = 1;
    for (int i = 0; i < m; ++i)                  // set MOD = 2^m
      MOD *= 2;
    ans = matPow(ans, n);                        // O(log n) 
    printf("%lld\n", ans.mat[0][1]);             // this if fib(n) % MOD
  }
  return 0;
}

File: ch5/UVa10229.py
import sys

MOD = 1

def mod(a, m):
  return (a%m+m)%m

def matMul(a, b):
  global MOD

  ans = [[0, 0], [0, 0]]
  for i in range(2):
    for k in range(2):
      if a[i][k] == 0:
        continue
      for j in range(2):
        ans[i][j] += mod(a[i][k], MOD) * mod(b[k][j], MOD)
        ans[i][j] = mod(ans[i][j], MOD)
  return ans

def matPow(base, p):
  ans = [[1, 0], [0, 1]]
  while p != 0:
    if p&1 != 0:
      ans = matMul(ans, base)
    base = matMul(base, base)
    p >>= 1
  return ans


if __name__ == '__main__':
  for line in sys.stdin:
    n, m = map(int, line.strip('\n').split())
    MOD = 2**m
    ans = matPow([[1, 1], [1, 0]], n)
    print(ans[0][1])
